### PR TITLE
Stats: Update Devices module with new empty state

### PR DESCRIPTION
**Pull request by grzegorz-cp**
Related to https://github.com/Automattic/red-team/issues/70

## Proposed Changes

* update the Devices module with a new empty state and place it behind a feature flag
* refactor TS definitions
* add the new skeleton loader to the module

## Why are these changes being made?

* a part of the empty states project

## Testing Instructions

* go to the live branch
* apply `stats/empty-module-traffic` feature flag
* verify that a page with an empty Devices component works with and without the feature flag
* repeat for blogs with data
* verify that the advanced feature overlay works properly
* verify that the new empty state only shows up after applying the feature flag

SCR-20240708-lhvl

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+	return (
+		<>
+			{ isNewEmptyStateEnabled && (
+				<>
+					{ showLoader && (
+						<StatsCardSkeleton
+							isLoading={ isFetching }
+							className={ className }
+							title={ devicesStrings.title }
+							type={ 3 }
```
Comment (kangzj) [inline client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: Not related to the PR: It missed my initial review, but I guess we could give the magic numbers some meaning here.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx [hunk 2]
```diff
@@ -0,0 +1,329 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
```
Comment (kangzj) [inline client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: The renaming of the file made it very hard to parse the changes. Can we please name it back and use a follow up to rename it please?

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+				<>
+					{ showLoader && (
+						<StatsCardSkeleton
+							isLoading={ isFetching }
+							className={ className }
+							title={ devicesStrings.title }
+							type={ 3 }
+						/>
+					) }
+					{ ! showLoader &&
```
Comment (kangzj) [inline client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: The code duplication from L167 to L258 is getting a bit out of control. Is there anything we could easily reuse with reasonable efforts here?

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 2 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: The name is the same. I think this code is shown as a new file because I added new code to the file to implement the empty state.
Is it helpful if you check the list of the commits? I moved the files first, without making any changes, to minimise the changes visible in the PR.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 3 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: I copied everything for two reasons: to make it easier to remove the future flag and to avoid the risk of leaking new changes. I believe that this is not only convenient but also safer.
Of course, we could minimise it, but future cleaning won't be as straightforward as removing a whole block of the old code.

I think we should keep it as it is here and focus on cleaning the future flags quickly, but if you feel strongly about it, I can work on it more to prioritise less code.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: I added a description for the property.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 2 above)
Comment (kangzj) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: I understand the file name is the same, but to git this is a name change. When doing renaming with not too much changes, git is clever enough to figure it out and keep the history of the file. But when it comes to lots of changes like this one, git determines they are not the same file anymore, meaning the revision would start fresh.

I could manually compare or use VSCode to compare two files, but anyway the review becomes harder. 

Based those two points, I would still suggest that we keep the file in the original folder for the PR, and then move it in a follow up if it makes sense.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 1 above)
Comment (kangzj) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: In this case, it doesn't seems to be important enough, but it is still confusing to have magic numbers. Not a blocker tho.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 3 above)
Comment (kangzj) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: I'm happy to ship and iterate 🙂

Just want to clarify that I was not talking about the branch for the feature flag. If you look at L167 to L258, two components using the same component `StatsCard`, which I suppose we created to unify the logic.

**Review by kangzj — APPROVED** [review]
Comments inlined but no blockers! 👍

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: I can change this in future updates. What do you think would be better? "OptionA" or "Variant1" as strings instead of numbers?

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 3 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: Lines 167 to 191 contain a new empty state. Lines 192 to 258 have the old component.
I'll update this to avoid repetition.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 2 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: ok, I will restore the files to the original directory in the main Stats folder.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 1 above)
Comment (kangzj) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: Either will be better than the magic numbers, with the naming we'd know they are just variants of the placeholder, not something else.

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 2 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: Done in 4397f93f19f42152c9b9f62e035fb2fecc550127 and 7367711fe2fd00dcb0f2fe37ef2903ca7730ab0e

File: client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx (hunk 3 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-module-devices/stats-module-devices.tsx]: Done in 633205f7546ca1f4b3bfb2050e11896d2ec47bb0

**Comment by grzegorz-cp** [comment]
I moved the files back @kangzj. I tested it but please test again because the new change updates the feature flag condition and affects many files with import paths.

**Review by kangzj — APPROVED** [review]
Looks good to me! Let's ship it 👍 

> Just want to clarify that I was not talking about the branch for the feature flag. If you look at L167 to L258, two components using the same component StatsCard, which I suppose we created to unify the logic.

I'd like to clarify again, I wasn't talking about the feature flag conditional branch, but I'm happy for the PR to ship anyway.

I was wondering - not just in this PR, `StatsCard` was only used as an empty state https://github.com/Automattic/wp-calypso/pull/92467/files#diff-3cf9bde5f0ad161211dd316095623b4b4190a75fad8275ba25a879cfc18be09dR169, and then following that, is the card displaying the data https://github.com/Automattic/wp-calypso/pull/92467/files#diff-3cf9bde5f0ad161211dd316095623b4b4190a75fad8275ba25a879cfc18be09dR201. Just wanted to point it out, so that we don't forget about it when refactoring.

Thanks for the changes. Looks so much easier to understand now! Happy to ship as is!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16253584

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @grzegorz-cp for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16253584).
